### PR TITLE
add retry logic to step-level

### PR DESCRIPTION
**Pull request by rescuerunaways**

### CODE REVIEW

File: cache/network/download.go [hunk 1]
```diff
@@ -48,12 +50,40 @@ func Download(ctx context.Context, params DownloadParams, logger log.Logger) (ma
...
-	downloadErr := downloadFile(ctx, retryableHTTPClient.StandardClient(), restoreResponse.URL, params.DownloadPath)
-	if downloadErr != nil {
-		return "", fmt.Errorf("failed to download archive: %w", downloadErr)
+	const maxRetries = 3
+	const retryDelay = 2 * time.Second
+	retriableErrors := []string{"Range request returned invalid Content-Length", "EOF"}
+
+	// Attempt to download, retrying on retriable errors
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		downloadErr := downloadFile(ctx, retryableHTTPClient.StandardClient(), restoreResponse.URL, params.DownloadPath)
```
Comment (viktorbenei) [inline cache/network/download.go:59]: We should check what happens if the target file already exists - and ensure that won't cause an issue (delete it if there's already a partially downloaded file if that'd cause an issue).

File: cache/network/download.go [hunk 2]
```diff
@@ -48,12 +50,40 @@ func Download(ctx context.Context, params DownloadParams, logger log.Logger) (ma
...
+			return "", fmt.Errorf("non-retriable error occurred: %w", downloadErr)
+		}
+
+		logger.Debugf("Retriable error occurred, attempt %d/%d: %v", attempt+1, maxRetries, downloadErr)
+
+		if ctx.Err() != nil {
+			return "", fmt.Errorf("context cancelled: %w", ctx.Err())
+		}
+		if attempt < maxRetries-1 {
+			time.Sleep(retryDelay)
```
Comment (viktorbenei) [inline cache/network/download.go:82]: What's the use case for waiting 2 seconds here?

File: cache/network/download.go [hunk 3]
```diff
@@ -48,12 +50,40 @@ func Download(ctx context.Context, params DownloadParams, logger log.Logger) (ma
 
 	logger.Debugf("Download archive")
 
-	downloadErr := downloadFile(ctx, retryableHTTPClient.StandardClient(), restoreResponse.URL, params.DownloadPath)
-	if downloadErr != nil {
-		return "", fmt.Errorf("failed to download archive: %w", downloadErr)
+	const maxRetries = 3
+	const retryDelay = 2 * time.Second
+	retriableErrors := []string{"Range request returned invalid Content-Length", "EOF"}
```
Comment (viktorbenei) [inline cache/network/download.go:55]: There's also `connection reset by peer` which might worth a retry.

Comment (viktorbenei) [reply cache/network/download.go:55]: also `read: operation timed out` might worth a retry - wdyt?

File: cache/network/download.go (hunk 2 above)
Comment (rescuerunaways) [reply cache/network/download.go:82]: it's a standard practice to give a network some chance to recover from whatever was an issue, 6 seconds total seems like a minimal possible burden; wdyt

File: cache/network/download.go [hunk 4]
```diff
@@ -48,12 +50,40 @@ func Download(ctx context.Context, params DownloadParams, logger log.Logger) (ma
...
+
+	// Attempt to download, retrying on retriable errors
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		downloadErr := downloadFile(ctx, retryableHTTPClient.StandardClient(), restoreResponse.URL, params.DownloadPath)
+		if downloadErr == nil {
+			return restoreResponse.MatchedKey, nil
+		}
+
+		isRetriable := false
+		for _, retriableError := range retriableErrors {
```
Comment (duplabe) [inline cache/network/download.go:65]: Could you add some tests? There is a lot of logic here for matching errors, handling context cancels, etc...

File: cache/network/download.go (hunk 1 above)
Comment (rescuerunaways) [reply cache/network/download.go:59]: I checked&tested, download is an idempotent operation
https://github.com/melbahja/got/blob/9c99581287dd94c9fceee95e5c9b502941903497/download.go#L201

File: cache/network/download.go (hunk 2 above)
Comment (rescuerunaways) [reply cache/network/download.go:83]: reduced to 1

**Review by duplabe — APPROVED** [review]

**Review by viktorbenei — APPROVED** [review]

**Pull request merged**